src/dune_pkg/rev_store.ml

```diff
@@ -23,11 +28,286 @@ module Object = struct
       && String.for_all s ~f:(function
         | '0' .. '9' | 'a' .. 'f' | 'A' .. 'F' -> true
         | _ -> false)
-    then Some (Sha1 (String.lowercase_ascii s))
+    then Some (String.lowercase_ascii s)
     else None
   ;;
 end
 
+module Commit = struct
+  module T = struct
+    type t =
+      { path : Path.Local.t
+      ; rev : Object.t
+      }
+
+    let compare { path; rev } t =
+      let open Ordering.O in
+      let= () = Path.Local.compare path t.path in
+      Object.compare rev t.rev
+    ;;
+
+    let to_dyn { path; rev } =
+      Dyn.record [ "path", Path.Local.to_dyn path; "rev", Object.to_dyn rev ]
+    ;;
+  end
+
+  include T
+  module C = Comparable.Make (T)
+  module Set = C.Set
+end
+
+module File = struct
+  module T = struct
+    type t =
+      | Redirect of
+          { path : Path.Local.t
+          ; to_ : t
+          }
+      | Direct of
+          { path : Path.Local.t
+          ; size : int
+          ; hash : string
+          }
+
+    let compare = Poly.compare
+
+    let to_dyn = function
+      | Redirect _ -> Dyn.opaque ()
+      | Direct { path; size; hash } ->
+        Dyn.record
+          [ "path", Path.Local.to_dyn path
+          ; "size", Dyn.int size
+          ; "hash", Dyn.string hash
+          ]
+    ;;
+  end
+
+  include T
+
+  let path = function
+    | Redirect p -> p.path
+    | Direct p -> p.path
+  ;;
+
+  let rec size = function
+    | Direct t -> t.size
+    | Redirect t -> size t.to_
+  ;;
+
+  let rec hash = function
+    | Direct t -> t.hash
+    | Redirect t -> hash t.to_
+  ;;
+
+  module C = Comparable.Make (T)
+  module Set = C.Set
+end
+
+module Cache = struct
+  let cache_dir =
+    lazy
+      (let path =
+         Path.L.relative
+           (Lazy.force Dune_util.xdg
+            |> Xdg.cache_dir
+            |> Path.Outside_build_dir.of_string
+            |> Path.outside_build_dir)
+           [ "dune"; "rev_store" ]
+       in
+       match Path.mkdir_p path with
+       | exception _ -> Error ()
+       | () -> Ok path)
+  ;;
+
+  (* Some operations on the cache may end up filling up the map. This helper
+     function allows us to give a more helpful error message in those cases. *)
+  let protect_map_full f =
+    match f () with
+    | x -> x
+    | exception Lmdb.Mdb.Map_full ->
+      let dir =
+        match Lazy.force cache_dir with
+        | Ok dir -> dir
+        | Error () ->
+          Code_error.raise "Map_full raised even though cache doesn't exist" []
+      in
+      User_error.raise
+        [ Pp.text "Revision store cache is full."; Pp.text "Please delete:"; Path.pp dir ]
+  ;;
+
+  let db =
+    lazy
+      (protect_map_full
+       @@ fun () ->
+       Lazy.force cache_dir
+       |> Result.map ~f:(fun path ->
+         Lmdb.Env.create
+           ~map_size:500_000_000 (* 500 MB *)
```

Comment:
Seems way too small. Let's add another 0 here at least.

Reply:
But remember lmdb is limited by address space, which on a 32bit architecture limits the db size to way less than 4GiB.

Reply:
@madroach do you know if lmdb has a way to automatically resize?